src/lib/operators/table_scan_implementation.hpp

```diff
@@ -0,0 +1,362 @@
+    if (value_segment) {
+      _process_segment(value_segment);
+      return;
+    }
+
+    throw std::logic_error("Unhandled segment type in _process_chunk");
+  }
+
+  // -----------------
+  // These methods go through the segment and add all relevant values to _chunk_offsets_to_add_to_result_table
+  // OR set _add_all_chunk_offsets to prevent copying the whole list
+  void _process_segment(std::shared_ptr<ValueSegment<T>> segment) {
+    const auto& values = segment->values();
+    const size_t values_size = values.size();
+
+    for (size_t value_index = 0; value_index < values_size; ++value_index) {
+      if (_scan_type_comparator(values[value_index])) {
+        _chunk_offsets_to_add_to_result_table.push_back(value_index);
+      }
+    }
+  }
+
+  void _process_segment(std::shared_ptr<DictionarySegment<T>> segment) {
+    const DictionarySegmentProcessingFlags flags(segment, _scan_type, _typed_search_value);
+
+    if (flags.add_none) {
+      return;
+    }
+
+    if (flags.add_all) {
+      _add_all_chunk_offsets = true;
+    } else {
+      const auto& attribute_vector = *(segment->attribute_vector());
+      size_t attribute_vector_size = attribute_vector.size();
+
+      for (size_t attribute_id = 0; attribute_id < attribute_vector_size; ++attribute_id) {
+        if (flags.should_add_value_id(attribute_vector.get(attribute_id))) {
+          _chunk_offsets_to_add_to_result_table.push_back(attribute_id);
+        }
+      }
+    }
+  }
+
+  void _process_segment(std::shared_ptr<ReferenceSegment> segment) {
+    ColumnID referenced_column_id = segment->referenced_column_id();
+    const std::shared_ptr<const Table> referenced_table = segment->referenced_table();
+    const auto& pos_list = (*segment->pos_list());
+
+    // Build a map: chunk_id to DictionarySegmentProcessingFlags (contains an entry if chunk at rowId.chunk_id
+    // is DictionarySegment). This is necessary so when we loop over all referenced values in the PosList,
+    // we can quickly decide whether an entry needs to be added to the result set.
+    // Also we decided to keep pointers to the relevant BaseAttributeVector and ValueSegments for each chunk_id
+    // so we do not always have to do dynamic casting with RTTI.
+
+    std::unordered_set<ChunkID> unique_referenced_chunk_ids;
+    for (const auto& rowId : pos_list) {
+      unique_referenced_chunk_ids.insert(rowId.chunk_id);
+    }
+
+    // initialize to true as we will and this together with some tests.
+    // (Can we win the price for the longest variable names for this?)
+    bool _all_referenced_segments_are_dictionary_segments_and_can_be_added_completely = true;
+
+    std::unordered_map<ChunkID, DictionarySegmentProcessingFlags> referenced_dictionary_segment_flags;
+    std::unordered_map<ChunkID, std::shared_ptr<const BaseAttributeVector>> referenced_dictionary_segment_vectors;
+    std::unordered_map<ChunkID, std::shared_ptr<ValueSegment<T>>> referenced_value_segments;
```

Comment:
It should be cheaper both in terms of memory usage and runtime costs to use vectors of `unique_referenced_chunk_ids.size()` length